### DISCUSSION TITLE
Service Graph does not show ingress-nginx to actual service requests

### DISCUSSION THREAD

**Discussion by chazapp** [grafana/tempo#3133] [Q&A] [answered]
Hello world,
I have configured all things GrafanaLabs in Terraform targeting a Minikube cluster. You can see the project here: [chazapp/o11y](https://github.com/chazapp/o11y)    
  
The cluster contains nginx-ingress and Grafana/Loki/Tempo/Prometheus/Grafana Agent/Pyroscope all configured and ready to go with one `terraform apply` command.  
  
I have also developped a React webapp instrumented with Faro connected to Grafana Agent and a Golang Gin API with GORM + Postgres database instrumented with OpenTelemetry SDKs from the code (_because Beyla wasn't being cooperative on Minikube on Windows 😄_). Both these are also applied via Terraform using their own Helm Charts   
  
I have configured OpenTelemetry in `nginx-ingress` to have a complete overview of my requests in Tempo, and my traces are looking great.  

![image](https://github.com/grafana/tempo/assets/15686688/349b2fb0-9c35-40fb-afff-ce7787345d52)
  

My issue is the Service Graph view of Tempo does not show the connections between `ingress-nginx` and the API. Am I doing something wrong or did I miss something ? I would have expected an arrow connecting these two nodes.   
  
![image](https://github.com/grafana/tempo/assets/15686688/7760107a-b0ff-402d-b4e3-690202942b4e)

  
Thank you for your help !

**Comment by joe-elliott** [comment, marked as answer]
Tempo looks for client/server pairs by using the "kind" field on the span. At a quick glance it looks like the nginx span is marked "server" as is the "wall-api" span. This would prevent the metrics generator from correctly pairing these up as a client/server pair.

Generally we would prefer not to attempt to match up every span for performance reasons, but perhaps that's necessary in situations like these to get an accurate service graph?
